crates/oxc_transformer/src/es2018/async_generator_functions/for_await.rs

```diff
@@ -0,0 +1,340 @@
+use oxc_semantic::{ScopeFlags, ScopeId, SymbolFlags};
+use oxc_span::SPAN;
+use oxc_traverse::{BoundIdentifier, TraverseCtx};
+
+use super::AsyncGeneratorFunctions;
+use crate::common::helper_loader::Helper;
+
+impl<'a, 'ctx> AsyncGeneratorFunctions<'a, 'ctx> {
+    pub(crate) fn transform_for_of_statement(
+        &mut self,
+        stmt: &mut ForOfStatement<'a>,
+        ctx: &mut TraverseCtx<'a>,
+    ) -> Vec<'a, Statement<'a>> {
+        let step_key =
+            ctx.generate_uid("step", ctx.current_scope_id(), SymbolFlags::FunctionScopedVariable);
+        // step.value
+        let step_value = ctx.ast.expression_member(ctx.ast.member_expression_static(
+            SPAN,
+            step_key.create_read_expression(ctx),
+            ctx.ast.identifier_name(SPAN, "value"),
+            false,
+        ));
+
+        let assignment_statement = match &mut stmt.left {
+            ForStatementLeft::VariableDeclaration(variable) => {
+                // for await (let i of test)
+                let mut declarator = variable.declarations.pop().unwrap();
+                declarator.init = Some(step_value);
+                let variable = ctx.ast.variable_declaration(
+                    SPAN,
+                    declarator.kind,
+                    ctx.ast.vec1(declarator),
+                    false,
+                );
+                let declaration = ctx.ast.declaration_from_variable(variable);
+                Statement::from(declaration)
+            }
+            left @ match_assignment_target!(ForStatementLeft) => {
+                // for await (i of test), for await ({ i } of test)
+                let target = ctx.ast.move_assignment_target(left.to_assignment_target_mut());
+                let expression = ctx.ast.expression_assignment(
+                    SPAN,
+                    AssignmentOperator::Assign,
+                    target,
+                    step_value,
+                );
+                ctx.ast.statement_expression(SPAN, expression)
+            }
+        };
+
+        let body = {
+            let mut statements = ctx.ast.vec_with_capacity(2);
+            statements.push(assignment_statement);
+            let stmt_body = &mut stmt.body;
+            // Skip empty BlockStatement
+            if matches!( stmt_body, Statement::BlockStatement(block) if !block.body.is_empty()) {
+                statements.push(ctx.ast.move_statement(stmt_body));
+            }
+            statements
+        };
+
+        build_for_await(
+            self.ctx.helper_load(Helper::AsyncIterator, ctx),
+            ctx.ast.move_expression(&mut stmt.right),
+            &step_key,
+            body,
+            ctx.current_scope_id(),
+            ctx,
+        )
+    }
+}
+
+/// Constructs AST based on AstBuilder
+/// <https://github.com/babel/babel/blob/d20b314c14533ab86351ecf6ca6b7296b66a57b3/packages/babel-plugin-transform-async-generator-functions/src/for-await.ts#L3-L30>
+fn build_for_await<'a>(
```

Comment:
Holy CHRIST! This function is *enormous*. We really do need the `ast!` macro.

Probably you were going to do this anyway, but could you include what code it produces in the comment (written out, rather than just link to Babel's impl)?

Reply:
Also, I'm not sure the 3 x `VariableDeclaration`s are actually needed. Doesn't Babel just discard them?

I had assumed they were only in Babel's template to make the `template` function resolve the identifiers correctly (I could be wrong).

Reply:
I think so, but anyway, we need to follow Babel's implementation 😢

Reply:
We need to follow Babel's *output*, but I don't think we need follow its *implementation* necessarily.

Anyway, I don't fully understand this transform, so I may be wrong in what I said above anyway.